tokio-executor/src/task/harness.rs

```diff
@@ -0,0 +1,592 @@
+                false
+            }
+            Ok(Poll::Pending) => {
+                let res = self.header().state.transition_to_idle();
+
+                if res.is_canceled() {
+                    self.do_cancel(res);
+                    false
+                } else {
+                    res.is_notified()
+                }
+            }
+            Err(err) => {
+                self.complete(executor, join_interest, Track::new(Err(Error::panic(err))));
+                false
+            }
+        }
+    }
+
+    pub(super) unsafe fn drop_task(mut self) {
+        let might_drop_join_waker_on_release = self.might_drop_join_waker_on_release();
+
+        // Read the join waker cell just to have it
+        let (join_waker, check) = self.read_join_waker();
+
+        // transition the task to released
+        let res = self.header().state.release_task();
+
+        assert!(res.is_complete() || res.is_canceled());
+
+        if might_drop_join_waker_on_release && !res.is_join_interested() {
+            debug_assert!(res.has_join_waker());
+
+            // Its our responsibility to drop the waker
+            check.check();
+            let _ = join_waker.assume_init();
+        }
+
+        if res.is_final_ref() {
+            self.dealloc();
+        }
+    }
+
+    unsafe fn dealloc(self) {
+        // Check causality
+        self.header().executor.with_mut(|_| {});
+        self.header().future_causality.with_mut(|_| {});
+        self.trailer().waker.with_mut(|_| {
+            // we can't check the contents of this cell as it is considered
+            // "uninitialized" data at this point.
+        });
+
+        // Drop task components. This is mostly to properly release loom's data
+        self.header.as_ptr().drop_in_place();
+        self.trailer.as_ptr().drop_in_place();
+
+        raw::dealloc::<T, _>(self.header);
+    }
+
+    // ===== join handle =====
+
+    pub(super) unsafe fn read_output(
+        mut self,
+        dst: *mut Track<super::Result<T::Output>>,
+        state: Snapshot,
+    ) {
+        let src = self.future_or_output as *const Track<super::Result<T::Output>>;
+
+        if state.is_canceled() {
+            dst.write(Track::new(Err(Error::cancelled())));
+        } else {
+            src.copy_to_nonoverlapping(dst, 1);
+        }
+
+        // Before transitioning the state, the waker must be read. It is
+        // possible that, after the transition, we are responsible for dropping
+        // the waker but before the waker can be read from the struct, the
+        // struct is deallocated.
+        let (waker, check) = self.read_join_waker();
+
+        // The operation counts as dropping the join handle
+        let res = self.header().state.complete_join_handle();
+
+        if res.is_released() {
+            // We are responsible for freeing the waker handle
+            check.check();
+            let _ = waker.assume_init();
```

Comment:
Nit/TIOLI: I think an explicit `drop` might express the intent slightly clearer here, since we're dropping the waker?
```suggestion
            drop(waker.assume_init());
```
up to you.